src/dump.c

```diff
@@ -1246,12 +1236,32 @@ static jl_value_t *jl_deserialize_datatype(ios_t *s, int pos, jl_value_t **loc)
             dt->alignment = MAX_ALIGN;
         dt->types = jl_emptysvec;
     }
-    dt->parameters = (jl_svec_t*)jl_deserialize_value(s, (jl_value_t**)&dt->parameters);
-    jl_gc_wb(dt, dt->parameters);
+
+    if (tag == 5) {
+        assert(pos > 0);
+        assert(mode != MODE_MODULE_POSTWORK);
+        arraylist_push(&flagref_list, loc);
+        arraylist_push(&flagref_list, (void*)(uintptr_t)pos);
+        dt->uid = -1; // mark that this type needs a new uid
+    }
+
+    int has_instance = (flags >> 3) & 1;
+    if (has_instance) {
+        assert(mode != MODE_MODULE_POSTWORK); // there shouldn't be an instance on a type with uid = 0
```

Comment:
also true, but this is asserting specifically that `dt->instance` isn't being assigned to